### PR TITLE
Code improvements with alpha model v2 + new ResNet34 model

### PR DESCRIPTION
**Pull request by Adelionek**
Co zawiera PR:
- Nowa metoda do ładowania danych
- Nowe metryki: ROC, Confusion Matrix
- Nowy eksperyment na modelu alfa z wykorzystaniem nowej metody ładowania danych i wykresami
- Szybka instrukcja w README.md
- Dodanie supportu dla device=MPS pod procesory M1/M2
- Drobne zmiany w metodzie do uczenia modelu + wyświetlanie bardziej szczegółowego progresu i metryk
- Nowy model ResNet34 z transfer learning - acc około 40%

Commity zawierają dodatkowe informacje o wprowadzonych zmianach.

### CODE REVIEW

File: data/loaders.py [hunk 1]
```diff
@@ -104,6 +115,51 @@ def load(self, verbose: bool = True):
         if verbose:
             print('Done.')
 
+
+    def load_v2(self, verbose: bool = True, subset_size: int = None):
```
Comment (emkarcinos) [inline data/loaders.py]: Wyrzućmy starą metodę `load` i zastąpmy ją tą nową

File: data/loaders.py [hunk 2]
```diff
@@ -104,6 +115,51 @@ def load(self, verbose: bool = True):
...
+        train_set_size_ratio = 0.7
```
Comment (emkarcinos) [inline data/loaders.py]: Może zwiększymy trochę zbiór treningowy? Ja bym zrobił ratio conajmniej 0.8 albo nawet 0.9

File: data/loaders.py [hunk 3]
```diff
@@ -104,6 +115,51 @@ def load(self, verbose: bool = True):
...
+        self.classes_count = len(full_dataset.classes)
+        self.label_names = dict([(value, key) for key, value in full_dataset.class_to_idx.items()])
+
+        if subset_size:
+            full_dataset = torch.utils.data.Subset(full_dataset, np.random.choice(len(full_dataset), subset_size, replace=False))
+    
+
+        train_size = int(train_set_size_ratio * len(full_dataset))
+        test_size = len(full_dataset) - train_size
+        train_dataset, test_dataset = torch.utils.data.random_split(full_dataset, [train_size, test_size])
```
Comment (emkarcinos) [inline data/loaders.py]: Nie wiem czy to jest najlepszy pomysł zeby te zbiory losować. Zależnie od tego jaki mamy seed możemy dostać różne rezultaty, wg mnie powinniśmy po prostu pierwsze x wartości brać jako trening a reszta jako test.

File: data/loaders.py [hunk 4]
```diff
@@ -104,6 +115,51 @@ def load(self, verbose: bool = True):
...
+
+        self.train_loader = torch.utils.data.DataLoader(
+            train_dataset,
+            batch_size=self.batch_size,
+            num_workers=4,
```
Comment (emkarcinos) [inline data/loaders.py]: Może to jakoś sparametryzować? Na przykład domyślnie wziąć liczbę corów obecnego CPU i podzielić przez 2, dorzucając do konstruktora opcje żeby to zcustomizować

File: data/loaders.py [hunk 5]
```diff
@@ -104,6 +115,51 @@ def load(self, verbose: bool = True):
...
+            shuffle=False
```
Comment (emkarcinos) [inline data/loaders.py]: Train można shufflować.

File: experiments/inference.py [hunk 6]
```diff
@@ -44,6 +46,48 @@ def evaluate(
     return trues, preds, losses
 
 
+def predict_image(
```
Comment (emkarcinos) [inline experiments/inference.py:49]: Zastąpiłbym wnętrze obecnej metody `infer` zamiast tworzyć nową.

File: experiments/inference.py [hunk 7]
```diff
@@ -44,6 +46,48 @@ def evaluate(
...
+    classifier: nn.Module,
+    classes: list[str],
+    device: str,
+    image
+) -> int:
+    """
+    Predict the class of image. Returns class index with top confidence.
+    """
+
+    classifier.to(device)
```
Comment (emkarcinos) [inline experiments/inference.py]: Czy na pewno chcemy robić predykcję na GPU? Dobrze byłoby sprawdzić czy dla pojedynczego obrazu nie jest szybciej zrobić to na CPU - transfer danych do GPU trochę zajmuje, a nie robimy tutaj żadnych obliczeń modyfikujących model więc powinno być szybko.

**Review by emkarcinos — DISMISSED** [review]
Kilka zmian.

Zaktualizowałbym obecny `model_alpha` o nowe metryki.

File: data/loaders.py (hunk 3 above)
Comment (Adelionek) [reply data/loaders.py]: Ja osobiście zostałbym przy tej metodzie, przewija się w  większości tutoriali przy ładowniu danych. Zgadzam się odnośnie seedu, w obecnej implementacji posplitowany dataset będzie niereprodukowalny. Do metody przekazałem obiekt generator z manual_seed(0).

File: data/loaders.py (hunk 2 above)
Comment (Adelionek) [reply data/loaders.py]: Zwiększone do 0.8 i sparametryzowne

File: data/loaders.py (hunk 1 above)
Comment (Adelionek) [reply data/loaders.py]: Done, wyrzuciłem też zbędną funkcje _load_data

File: data/loaders.py (hunk 4 above)
Comment (Adelionek) [reply data/loaders.py]: Bardzo dobra sugestia - zmienione

File: data/loaders.py (hunk 5 above)
Comment (Adelionek) [reply data/loaders.py]: Właśnie nie mogłem się zdecydować, poprzednie commity miały True ale jednak postawiłem na False 😄  Zmienione na True

File: experiments/inference.py (hunk 6 above)
Comment (Adelionek) [reply experiments/inference.py:49]: To trochę inne metody. 

`infer` to mała prosta metoda. Jest wykorzystywana do treningu, liczenia lossu i predykcji.

`predict_image` jest bardziej rozbudowana, narzuca transformacje na dane tak aby odwzorować te z testSetu (możemy użyć img z zewnątrz naszego datasetu). Dodatkowo ta metoda zwraca index przewidzianej klasy i printuje top3 klasy. Wrzucenie tej logiki do infer przeciąży tą metodę, może też się to odbić na performance przy uczeniu bo ta metoda jest wywoływana bardzo dużo razy (przy każdym batchu).

File: experiments/inference.py (hunk 7 above)
Comment (Adelionek) [reply experiments/inference.py]: Trochę .... nanosekund to zajmuje 😄 
image

Spoczko - zmienione na CPU 👍

**Review by emkarcinos — COMMENTED** [review]
Spoko poprawki.

Mam jeszcze jedną małą sugestię co do tych jupyterów: Teraz będą trzy praktycznie takie same jupytery a mógłby być jeden:
- wspólny kod wyabstrachowany do modułow pythonowych
- różnice między jupyterami można po prostu dorzucić jako dodatkowe cele

Co sądzisz?

**Comment by Adelionek** [comment]
Jak najbardziej. Docelowo trzeba przenieść ten kod który się powtarza  do zewnętrznych modułów. 
Dwa pytania:
 - Czy robić ten refactoring w tym pull requeście?
 - Czy chcemy faktycznie wrzucać kilka modeli do jednego jupytera? W niektórych przypadkach myślę że chcielibyśmy mieć 1 eksperyment i 1 model per 1 plik. Jak to wyizolujemy to będziemy mogli sprawnie uczyć wiele różnych modeli na wspólnych interfejsach generując spójne metryki. Łatwiej też będzie z tego zrobić skrypt żeby przekazać paramsy.  Z drugiej strony jak będziemy mieli 1 wspólny plik to zawsze można to rozbić w razie potrzeby.

**Comment by emkarcinos** [comment]
Spoko, możemy zrobić w innym issue

**Comment by Adelionek** [comment]
@emkarcinos Coś źle kliknąłem przy mergowaniu i chyba twój approval się zresetował o ile był

**Review by emkarcinos — APPROVED** [review]
